Added an option to change locations of the MPI buffers for the communication layers

ravil-mobile opened the pull request:
This PR adds an option to perform MPI p2p non-blocking communication using either managed (unified), or host, or device memory. Some MPI GPU-aware libraries have problems with sending data over the unified memory (either too slow or just not capable of doing so e.g., CrayMPI v8.1.23)

Note: this PR depends on: https://github.com/SeisSol/Device/pull/23

uphoffc commented:
Hi Ravil,

this looks great. Two general comments:
1) I would not call this a "prefetch". Here, you introduce MPI-through-host via explicit copies. So the name of the feature should suggest that you work-around a bad or non-existent GPU-aware MPI implementation. Other codes such a flag "have GPU-aware mpi" or similar.
2) From a software design perspective, I would not introduce yet another ifdef, because ifdefs tend to code that does not compile, because it is not regularly checked. (And if you test it in CI then you increase the cost of CI by yet another build.) Here, you already have most of the things you need in order to make this a parameter or an environment variable. The AbstractCommunicationManager already offers an abstract interface, so you when you create the ghost cluster you can choose at run-time whether you want to create a "generic" ghost time cluster or one with "prefetch". The same holds true for the additional memory that is allocated: You can check at run-time whether it needs to be allocated or whether it can be omitted. I think it might be best to allocate the additional memory inside of GhostTimeClusterWithPrefetch because I don't think you need it anywhere else.

Best regards,
Carsten

ravil-mobile commented:
Hi @uphoffc , thanks for your comments.

>I would not call this a "prefetch". Here, you introduce MPI-through-host via explicit copies. So the name of the feature should suggest that you work-around a bad or non-existent GPU-aware MPI implementation. Other codes such a flag "have GPU-aware mpi" or similar.

It is a good point. However, in stead of just copying the data I am trying to copy them asynchronously, potentially overlapping with other tasks (for example, computations of the interior LTS layers).

Taking into account the MPI user-buffers are allocated with the unified/managed memory, there are two possible cases, namely: 1) copy them to the device memory (should be fast) and 2) copy to the host memory (usually slow).

The first case can be handled with `USE_GPU_AWARE_MPI=OFF`. What would you suggest for the second case? How shall I name it?

uphoffc commented:
Maybe in that case an enum would be self-explaining. E.g.
```c++
enum class GpuMPIMode {
  Direct,                            /// Direct MPI
  CopyInCopyOutDevice, /// Copy-in / copy-out to device memory
  CopyInCopyOutHost     /// Copy-in / copy-out to host memory
};
```

ravil-mobile commented:
> Maybe in that case an enum would be self-explaining. E.g.
> 
> ```
> enum class GpuMPIMode {
>   Direct,                            /// Direct MPI
>   CopyInCopyOutDevice, /// Copy-in / copy-out to device memory
>   CopyInCopyOutHost     /// Copy-in / copy-out to host memory
> };
> ```

Thanks! God you.

Btw, do you know by any chance whether the following lines of code can result in sending zero-size messages?

https://github.com/SeisSol/SeisSol/blob/f8f5a948ba1978cd6ee8388756a64bf477538421/src/Solver/time_stepping/GhostTimeCluster.cpp#L13-L20
https://github.com/SeisSol/SeisSol/blob/f8f5a948ba1978cd6ee8388756a64bf477538421/src/Solver/time_stepping/GhostTimeCluster.cpp#L29-L36

I mean that `meshStructure->neighboringClusters[region][0]` could be equal to zero for some reason? or we filter such cases during pre-processing?

uphoffc commented:
A partition might have empty time clusters. E.g. if the minimum time-step is 4*dt and the max timestep is 16*dt, then there might be no elements with 8*dt. Therefore, there might be empty messages. I don't recall any piece of code that would filter out these messages, but I am not 100% sure.

krenzland commented:
Yup, I think it probably does send empty messages. Be aware that changing this could lead to bugs in the synchronization between clusters as it could lead to unmatched receives...

ravil-mobile commented:
Thanks @uphoffc  and @krenzland. I will keep in my mind.

@uphoffc , could you, please, have a look at this PR one more time? I removed if-defs as you suggested.

Currently, I added the env. variable. The ideal case would be to add this option to the parameter-file. I think it is better to do once PR #829 is merged.

krenzland commented on src/Solver/time_stepping/GenericGhostTimeCluster.h:
```diff
@@ -6,23 +6,22 @@
-
-class GhostTimeCluster : public AbstractTimeCluster {
- private:
+class GenericGhostTimeCluster : public AbstractTimeCluster {
```
IMO the design should be different. Both GhostTimeClusters should inherit from AbstractGhostTimeCluster, which defines the interface.
This here should be called DirectGhostTimeCluster or simular.
testQueue can be a non-class function (poor design on my part there!)

ravil-mobile replied:
> Both GhostTimeClusters should inherit from AbstractGhostTimeCluster, which defines the interface.

Done. I moved all common parts to `AbstractGhostTimeCluster`. Both `DirectGhostTimeCluster` and `DirectGhostTimeCluster` add their own behavior i.e., `sendCopyLayer`, `receiveGhostLayer` and `testForGhostLayerReceives`

> testQueue can be a non-class function
I don't know what to do with it. As for me, it is not super critical here.

krenzland commented on src/Solver/time_stepping/GhostTimeClusterWithCopy.h:
```diff
@@ -0,0 +1,41 @@
+  std::vector<void*> prefetchCopyRegionsStreams{};
+  std::vector<void*> prefetchGhostRegionsStreams{};
+
+  enum ReceiveState { RequiresMpiTesting, RequiresPrefetchTesting, Ready };
```
should be enum class

ravil-mobile replied:
done

krenzland commented on src/Solver/time_stepping/GhostTimeClusterWithCopy.cpp:
```diff
@@ -0,0 +1,199 @@
+  lastSendTime = ct.correctionTime;
+  auto prefetchedRegions = prefetchCopyLayer();
+
+  while (not prefetchedRegions.empty()) {
```
I know that this is legal C++ but it's kinda weird :D

ravil-mobile replied:
Ok. I will change this to `!prefetchedRegions.empty()`

krenzland commented on src/Solver/time_stepping/GhostTimeClusterWithCopy.cpp:
```diff
@@ -0,0 +1,199 @@
+template <MPI::DataTransferMode CommType>
+void GhostTimeClusterWithCopy<CommType>::receiveGhostLayer() {
+  SCOREP_USER_REGION("receiveGhostLayer", SCOREP_USER_REGION_TYPE_FUNCTION)
+  assert(ct.predictionTime > lastSendTime);
```
You can change this to >=, there's an open issue somewhere that this assert is buggy

ravil-mobile replied:
Ok. I changed in both places, namely: here and in the parent class.

krenzland commented on src/Solver/time_stepping/GhostTimeClusterWithCopy.cpp:
```diff
@@ -0,0 +1,199 @@
+      auto* stream = prefetchGhostRegionsStreams[*region];
+      if (device.api->isStreamWorkDone(stream)) {
+        receiveRegionsStates[*region] = Ready;
+        // Note: fall to `Ready` is intentional
```
I don't get this comment

krenzland replied:
Ah, you mean fallthrough to ready is intentional? Yeah, makes sense.

ravil-mobile replied:
yes, `[[fallthrough]]` only works inside `switch`-statements. I just edited the comment

krenzland reviewed (changes requested):
Looks mostly good! The design overall is nice, however I think it may be a good idea to introduce an AbstractGhostTimeCluster (see comment in GenericGhostTimeCluster.h

ravil-mobile commented on src/Solver/time_stepping/TimeManager.cpp:
```diff
@@ -378,3 +382,7 @@ void seissol::time_stepping::TimeManager::setTv(double tv) {
   }
 }
 
+void seissol::time_stepping::TimeManager::releaseCommunicationManager() {
```
ok. done

ravil-mobile commented on src/Parallel/MPI.cpp:
```diff
@@ -90,4 +92,36 @@ void seissol::MPI::bindAcceleratorDevice() {
 }
 #endif
 
+void seissol::MPI::checkEnvVariables() {
```
done

uphoffc commented on src/SeisSol.cpp:
```diff
@@ -184,6 +186,8 @@ void seissol::SeisSol::finalize()
 
 	const int rank = MPI::mpi.rank();
 
+	m_timeManager.resetCommunicationManager();
+
```
Not sure why this is necessary.

uphoffc replied:
Ah that makes sense. But why don't you call it freeCommunicationManager if that's the purpose?

uphoffc commented on src/Solver/time_stepping/AbstractGhostTimeCluster.h:
```diff
@@ -1,28 +1,27 @@
-#ifndef SEISSOL_GHOSTTIMECLUSTER_H
-#define SEISSOL_GHOSTTIMECLUSTER_H
+#pragma once
```
pragma once is non-standard

uphoffc commented on src/Solver/time_stepping/DirectGhostTimeCluster.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef SEISSOL_GHOSTTIMECLUSTER_H
+#define SEISSOL_GHOSTTIMECLUSTER_H
```
Headerguard should have same name as file

ravil-mobile replied:
I replaced to `#pragma once`. If you and @krenzland make a decision to keep using header guards then I will revert and rename it

uphoffc commented on src/Solver/time_stepping/GhostTimeClusterFactory.h:
```diff
@@ -0,0 +1,52 @@
+#pragma once
```
pragma once is non-standard

krenzland replied:
Yes but supported by nearly everything: https://en.wikipedia.org/wiki/Pragma_once#Portability

ravil-mobile replied:
@uphoffc, @krenzland, Shall I leave `#pragma once` or replace it with the guards? 

I personally agree with @krenzland, this pragma is supported by almost all compilers and there will be no issues with it.

uphoffc replied:
I don't mind, that's for you guys to decide.

uphoffc commented on src/Solver/time_stepping/GhostTimeClusterWithCopy.h:
```diff
@@ -0,0 +1,40 @@
+#pragma once
```
pragma once is non-standard

uphoffc commented on src/Solver/time_stepping/GhostTimeClusterWithCopy.cpp:
```diff
@@ -0,0 +1,199 @@
+  }
+}
+
+template <MPI::DataTransferMode CommType>
+GhostTimeClusterWithCopy<CommType>::~GhostTimeClusterWithCopy() {
+  for (size_t region = 0; region < numberOfRegions; ++region) {
+    device.api->destroyGenericStream(prefetchCopyRegionsStreams[region]);
+    device.api->destroyGenericStream(prefetchGhostRegionsStreams[region]);
+    if constexpr (CommType == MPI::DataTransferMode::CopyInCopyOutDevice) {
+      device.api->freeMem(duplicatedCopyRegions[region]);
+      device.api->freeMem(duplicatedGhostRegions[region]);
+    } else {
+      device.api->freePinnedMem(duplicatedCopyRegions[region]);
+      device.api->freePinnedMem(duplicatedGhostRegions[region]);
+    }
+  }
+}
+
```
If you have a custom dtor you should also have custom copy / move operators and copy-/move-initializers. If you do not plan on copying / moving objects of this class then delete the 4 operators.

ravil-mobile replied:
I deleted them

uphoffc commented on src/Solver/time_stepping/TimeManager.cpp:
```diff
@@ -378,3 +382,7 @@ void seissol::time_stepping::TimeManager::setTv(double tv) {
   }
 }
 
+void seissol::time_stepping::TimeManager::resetCommunicationManager() {
+  communicationManager.reset(nullptr);
+}
```
reset accepts a double and not a pointer. Also not sure why this function is neccessary.

ravil-mobile replied:
Hi @uphoffc , I didn't get why it should accept `double`. `communicationManager` is a unique pointer. Calling reset will trigger the destructor of the underlying object.

Note, the `SeisSol` object is static and thus it is going to be deleted only at the very very end. At this point, the connection with the GPU driver will be lost. Omitting a call to the dtor of `communicationManager` will trigger a segfault because `WithCopy` implementations uses device memory which is freed by the dtor

uphoffc replied:
Ok sorry about that. I saw that AbstractCommunicationManager has a reset function and did not notice that it is stored as unique_ptr in TimeManager.

ravil-mobile commented on src/SeisSol.cpp:
```diff
@@ -95,9 +95,11 @@ bool seissol::SeisSol::init(int argc, char* argv[]) {
   logInfo(rank) << "Using MPI with #ranks:" << MPI::mpi.size();
+  // TODO (Ravil, David): switch to reading MPI options from the parameter-file.
+  MPI::mpi.checkEnvVariables();
 #endif
 #ifdef _OPENMP
-  pinning.checkEnvVariables();
+  pinning.setDataTransferModeFromEnv();
```
@uphoffc, thanks for noticing. Done

ravil-mobile commented on Documentation/gpus.rst:
```diff
@@ -118,3 +118,15 @@ the following will force SeisSol to allocate 1.5GB of stack GPU memory for tempo
+
+Currently, SeisSol allocates MPI user-buffers using the unified/managed memory
+type. Some MPI implementations perform poorly during non-blocking
+point-to-point communication. SeisSol provides *SEISSOL_PREFERRED_MPI_DATA_TRANSFER_MODE*
```
Done

ravil-mobile commented on Documentation/gpus.rst:
```diff
@@ -118,3 +118,15 @@ the following will force SeisSol to allocate 1.5GB of stack GPU memory for tempo
+will result in utilizing the regular device memory for non-blocking
+communication. The default value is *direct* which means that the communication
+goes over the unified/managed memory and thus does not involve explicit data
+copies.
```
Added. See here: https://github.com/SeisSol/SeisSol/blob/ravil/comm-layer/Documentation/LatexFigures/gpu-comm-layer-data-flow.png

uphoffc replied:
Great figure, thanks!

uphoffc commented on cmake/process_users_input.cmake:
```diff
@@ -146,9 +146,9 @@ if (NOT ${DEVICE_ARCH} STREQUAL "none")
 
     if (${DEVICE_ARCH} MATCHES "sm_*")
-        set(ALIGNMENT  64)
-    elseif(${DEVICE_ARCH} MATCHES "gfx*")
         set(ALIGNMENT  128)
+    elseif(${DEVICE_ARCH} MATCHES "gfx*")
+        set(ALIGNMENT  256)
```
ok

uphoffc reviewed (approved):
I suggest to rename resetCommunicationManager to freeCommunicationManager in TimeManager. LGTM otherwise.

krenzland commented:
Yup, I agree with Carsten here, otherwise it looks great!
